File: project_euler_py/problem_017.py
"""If the numbers 1 to 5 are written out in words: one, two, three,
four, five, then there are 3 + 3 + 5 + 4 + 4 = 19 letters used in total.

If all the numbers from 1 to 1000 (one thousand) inclusive were written
out in words, how many letters would be used?

Note: Do not count spaces or hyphens. For example,
342 (three hundred and forty-two) contains 23 letters and
115 (one hundred and fifteen) contains 20 letters.
The use of "and" when writing out numbers is in compliance with British usage.
"""
from math import log10
import sys

class TailRecurseException(BaseException):
  def __init__(self, args, kwargs):
    self.args = args
    self.kwargs = kwargs

def tail_call_optimized(g):
  """
  This function decorates a function with tail call
  optimization. It does this by throwing an exception
  if it is it's own grandparent, and catching such
  exceptions to fake the tail call optimization.
  
  This function fails if the decorated
  function recurses in a non-tail context.
  """
  def func(*args, **kwargs):
    f = sys._getframe()
    if f.f_back and f.f_back.f_back \
        and f.f_back.f_back.f_code == f.f_code:
      raise TailRecurseException(args, kwargs)
    else:
      while 1:
        try:
          return g(*args, **kwargs)
        except TailRecurseException as e:
          args = e.args
          kwargs = e.kwargs
  func.__doc__ = g.__doc__
  return func

@tail_call_optimized
def in_words(input_num):
    num_to_word = {
        1000: 'thousand',
        100: 'hundred',
        90: 'ninety',
        80: 'eighty',
        70: 'seventy',
        60: 'sixty',
        50: 'fifty',
        40: 'forty',
        30: 'thirty',
        20: 'twenty',
        19: 'nineteen',
        18: 'eighteen',
        17: 'seventeen', 
        16: 'sixteen',
        15: 'fifteen',
        14: 'fourteen',
        13: 'thirteen',
        12: 'twelve',
        11: 'eleven',
        10: 'ten',
        9: 'nine',
        8: 'eight',
        7: 'seven',
        6: 'six',
        5: 'five',
        4: 'four',
        3: 'three',
        2: 'two',
        1: 'one'
    }

    out = ''
    for num, name in num_to_word.items():
        if input_num == 0:
            return out

        elif input_num < 10 and input_num/num > 0:
            return out + f'{name}'

        elif input_num < 100 and input_num/num > 0:
            if input_num / num == 0:
                return out + f'{name}'
            else:
                return out + f'{name} ' + in_words(input_num % num)

        elif input_num % 100 == 0 and input_num/num > 0:
            return out + in_words(input_num/num) + f'{name}'
            + in_words(input_num % num)

        elif input_num/num > 0:
            return out + in_words(input_num/num) + f'{name} and ' + in_words(input_num % num)
    

print(in_words(150))

File: project_euler_rb/problem_017.rb
# If the numbers 1 to 5 are written out in words: one, two, three, four, five, then there are: 
# 3 + 3 + 5 + 4 + 4 = 19 letters used in total.

# If all the numbers from 1 to 1000 (one thousand) inclusive were written out in words, how many letters would be used?

# NOTE: Do not count spaces or hyphens.
# For example, 342 (three hundred and forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20 letters. 
# The use of "and" when writing out numbers is in compliance with British usage.

def in_words(int)
  num_to_word  = {
    1000 => 'thousand',
    100 => 'hundred',
    90 => 'ninety',
    80 => 'eighty',
    70 => 'seventy',
    60 => 'sixty',
    50 => 'fifty',
    40 => 'forty',
    30 => 'thirty',
    20 => 'twenty',
    19 => 'nineteen',
    18 => 'eighteen',
    17 => 'seventeen', 
    16 => 'sixteen',
    15 => 'fifteen',
    14 => 'fourteen',
    13 => 'thirteen',
    12 => 'twelve',
    11 => 'eleven',
    10 => 'ten',
    9 => 'nine',
    8 => 'eight',
    7 => 'seven',
    6 => 'six',
    5 => 'five',
    4 => 'four',
    3 => 'three',
    2 => 'two',
    1 => 'one'
  }

  str = ''
  num_to_word.each do |num, name|

    if int == 0
      return str

    elsif int < 10 == 1 && int/num > 0
      return str + "#{name}"

    elsif int < 100 && int/num > 0
      return str + "#{name}" if int%num == 0
      return str + "#{name} " + in_words(int%num)

    elsif int % 100 == 0 && int/num > 0
      return str + in_words(int/num) + " #{name}" + in_words(int % num)

    elsif int/num > 0
      return str + in_words(int/num) + " #{name} and " + in_words(int%num)

    end
  end
end

a = 1.upto(1000).
  map{ |int| in_words(int) }.
  map{ |word| word.gsub(' ', '').split('') }.
  flatten.
  length

puts a

# Complete!
# Answer = 21124
